wgsl/scanner.cc

```diff
@@ -0,0 +1,927 @@
+
+#include <tree_sitter/parser.h>
+#include <wctype.h>
+#include <algorithm>
+#include <bitset>
+#include <cassert>
+#include <cstring>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#if 0
+#define LOG(msg, ...) printf(msg "\n", ##__VA_ARGS__)
+#else
+#define LOG(...)
+#endif
+
+namespace {
+
+/// The possible external tokens matched by this custom scanner.
+/// The order of the entries in this enumerator must match the 'externals' in
+/// the grammar.js.
+enum Token {
+  BLOCK_COMMENT,
+  DISAMBIGUATE_TEMPLATE,  // A zero-length token used to scan ahead
+  TEMPLATE_ARGS_START,
+  TEMPLATE_ARGS_END,
+  LESS_THAN,           // '<'
+  LESS_THAN_EQUAL,     // '<='
+  SHIFT_LEFT,          // '<<'
+  SHIFT_LEFT_ASSIGN,   // '<<='
+  GREATER_THAN,        // '>'
+  GREATER_THAN_EQUAL,  // '>='
+  SHIFT_RIGHT,         // '>>'
+  SHIFT_RIGHT_ASSIGN,  // '>>='
+
+  // Special value, found found in grammar.js.
```

Comment:
Comment on my own code: This line makes no sense and should be removed. The URL is good though.